File: apps-microservices/graph-rag-api-recherche-rust-service/src/routers/fournisseur.rs
use actix_web::{get, web, HttpResponse};
use serde_json::json;

use crate::services::fournisseur_service::FournisseurService;

/// GET /fournisseur/{id_fournisseur} —
/// Get geographic coverage by fournisseur ID.
#[utoipa::path(
    get,
    path = "/fournisseur/{id_fournisseur}",
    params(("id_fournisseur" = String, Path, description = "Fournisseur ID")),
    responses(
        (status = 200, description = "Fournisseur geographic coverage")
    ),
    tag = "Fournisseur"
)]
#[get("/fournisseur/{id_fournisseur}")]
pub async fn get_couverture_by_fournisseur(
    path: web::Path<String>,
) -> HttpResponse {
    let id_fournisseur = path.into_inner();
    match FournisseurService::get_couverture_by_fournisseur(&id_fournisseur).await {
        Some(response) => HttpResponse::Ok().json(response),
        None => HttpResponse::NotFound().json(json!({
            "detail": format!("Fournisseur with ID '{}' not found.", id_fournisseur)
        })),
    }
}

/// GET /fournisseur/produit/{id_produit} —
/// Get geographic coverage by product ID.
#[utoipa::path(
    get,
    path = "/fournisseur/produit/{id_produit}",
    params(("id_produit" = String, Path, description = "Product ID")),
    responses(
        (status = 200, description = "Fournisseur geographic coverage for product")
    ),
    tag = "Fournisseur"
)]
#[get("/fournisseur/produit/{id_produit}")]
pub async fn get_couverture_by_produit(
    path: web::Path<String>,
) -> HttpResponse {
    let id_produit = path.into_inner();
    match FournisseurService::get_couverture_by_produit(&id_produit).await {
        Some(response) => HttpResponse::Ok().json(response),
        None => HttpResponse::NotFound().json(json!({
            "detail": format!("Geo coverage not found for product '{}'.", id_produit)
        })),
    }
}

File: apps-microservices/graph-rag-api-recherche-optim-service/app/routers/fournisseur.py
from fastapi import APIRouter, HTTPException, Path, status
import logging
from typing import Optional

from app.domain.models import FournisseurGeoResponse
from app.services.fournisseur_service import fournisseur_service

router = APIRouter()


@router.get(
    "/{id_fournisseur}",
    response_model=FournisseurGeoResponse,
    description="Get the geographical coverage (Pays and Departements) for a specific supplier.",
)
async def get_fournisseur_geo_coverage(
    id_fournisseur: str = Path(..., description="The ID of the supplier")
):
    try:
        geo_coverage = await fournisseur_service.get_geo_coverage(id_fournisseur)

        if not geo_coverage:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Fournisseur with ID '{id_fournisseur}' not found.",
            )

        return geo_coverage

    except HTTPException:
        raise
    except Exception as e:
        logging.error(f"Error fetching geo coverage for {id_fournisseur}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )


@router.get(
    "/produit/{id_produit}",
    response_model=FournisseurGeoResponse,
    description="Get the geographical coverage for a supplier associated with a specific product ID.",
)
async def get_geo_coverage_by_produit_id(
    id_produit: str = Path(..., description="The ID of the product")
):
    try:
        geo_coverage = await fournisseur_service.get_geo_coverage_by_produit(id_produit)

        if not geo_coverage:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Geo coverage not found for product '{id_produit}' (Product or Supplier might be missing).",
            )

        return geo_coverage

    except HTTPException:
        raise
    except Exception as e:
        logging.error(f"Error fetching geo coverage for product {id_produit}: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Internal server error",
        )
